internal/backends/mcu/renderer.rs

```diff
@@ -22,82 +21,130 @@ use i_slint_core::{Color, Coord, ImageInner, StaticTextures};
 
 type DirtyRegion = PhysicalRect;
 
-pub fn render_window_frame(
+pub trait LineBufferProvider {
```

Comment:
I couldn't just use a FnMut because of the `set_dirty_region` function.

Also I realize that this design will not support drawing the line from different cores. And also do not allow for any kind of asynchronous rendering.

Reply:
> Also I realize that this design will not support drawing the line from different cores. And also do not allow for any kind of asynchronous rendering.

Why not? Couldn't the implementation of process_line() call render_fn(), then set up the dma transfer and return. And the first thing the function would do is wait for an available buffer before calling render_fn()?

I agree about the multi-core part though. I think we should cross that bridge when we get there though :)

Reply:
> Why not? Couldn't the implementation of process_line() call render_fn(), then set up the dma transfer and return. And the first thing the function would do is wait for an available buffer before calling render_fn()?

Yes, that works for DMA, but this design don't allow the user code to run in a priority 0 for example.
And we also can't do asynchronous rendering if some operation within the lines could be done directly with DMA  (eg: blending)
Since we have like many DMA channel, we could imagine rendering into something like 4 lines in parallel, and have the CPU still do some work at the same time.

But this design doesn't allow that.

Reply:
You're right, the design doesn't allow for that.

My guts feeling is that we're more likely going to prioritise implementing support for something like async DMA2D or multi-core rendering on a higher-end device that has more RAM and uses frame buffers than on a device like the pico. In other words: I think the acceleration support will be much more important on the frame buffer renderer than on the line-by-line renderer.